### PR TITLE
feat(claude-code): report first-level subagent traces

### PR DESCRIPTION
**Pull request by zhouxiaohuan**
## Summary

This PR adds first-level subagent trace reporting for Claude Code.

- correlate `Agent` tool results with their generated subagent transcripts
- export direct subagent LLM and tool records into the parent turn trace
- emit `Agent` tool call/result spans instead of silently skipping them
- preserve the parent trace, session, and turn context for correlated records
- annotate subagent records with explicit agent hierarchy attributes
- merge parent and subagent records in timestamp order

## Motivation

Claude Code stores work delegated through the `Agent` tool in separate subagent transcript files.

Previously, these transcripts were not exported, and `Agent` tool spans were explicitly skipped. As a result, delegated LLM calls and tool executions were invisible in traces, producing unexplained latency gaps in the parent request.

This change makes direct subagent activity observable as part of the parent trace.

### Transcript correlation

The transcript parser now extracts `agentId` and `agentType` from the `Agent` tool result and associates them with the corresponding `tool_use_id`.

The hook processor then resolves the direct subagent transcript from:

```text
<parent-session>/subagents/agent-<agent-id>.jsonl
```

Only the basename of the agent identifier is used when resolving the path.

### Trace propagation

Subagent records inherit the parent:

- `trace_id`
- `gen_ai.session.id`
- `gen_ai.turn.id`

The following attributes identify the subagent relationship:

```text
gen_ai.agent.scope = subagent
gen_ai.agent.depth = 1
gen_ai.agent.id
gen_ai.agent.name
gen_ai.agent.parent.id
gen_ai.subagent.parent_tool_call.id
```

### Depth policy

Only direct, first-level subagent transcripts are expanded.

Nested `Agent` calls inside a subagent remain visible as tool spans, but their own transcripts are not recursively expanded. This keeps trace volume and traversal behavior bounded.

### Fail-open behavior

If a subagent transcript is absent or cannot be resolved, the parent trace continues to be exported unchanged.

No additional hook events, configuration, API, or storage schema changes are introduced.

## Compatibility

- parent-only Claude Code sessions retain their existing behavior
- transcript offset advancement remains owned by the parent session
- export remains bound to the existing Stop hook flow
- existing masking and content policies are applied to subagent records
- no changes are required for other supported agents

## Validation

- `npm run typecheck`
- `npm test`
- `npm run build`
- verified package generation with Node.js 22
- verified the packaged Pilot starts successfully in a Linux systemd environment

## Scope

This PR intentionally supports first-level subagents only. Recursive expansion of deeper subagent transcripts is out of scope.

### CODE REVIEW

File: assets/hooks/claude-code-hook-processor.mjs [hunk 1]
```diff
@@ -455,11 +492,65 @@ async function exportSession(state, stopReason) {
...
+        if (!childTranscriptPath || !fs.existsSync(childTranscriptPath)) continue;
+
+        const childParseResult = parseClaudeTranscript(childTranscriptPath, 0);
+        let childHash = INITIAL_HASH;
+        for (let childTurnIndex = 0; childTurnIndex < childParseResult.turns.length; childTurnIndex++) {
+          const childBuild = buildTurnRecords(
+            childParseResult.turns[childTurnIndex],
+            childTurnIndex,
+            link.agentId,
+            childHash,
```
Comment (ralf0131) [inline assets/hooks/claude-code-hook-processor.mjs]: **[Warning]** Consider adding a `try/catch` around `parseClaudeTranscript(childTranscriptPath, 0)`. If the subagent transcript file exists but contains malformed JSON or an unexpected schema, an uncaught exception here would abort the entire `exportSession` for the parent session. A per-subagent catch with a warning log would preserve the fail-open contract described in the PR body.

File: assets/hooks/claude-code-hook-processor.mjs [hunk 2]
```diff
@@ -85,6 +85,43 @@ function defaultLogDir() {
   return path.join(pilotDataDir(), 'logs', AGENT_ID);
 }
 
+function isAgentTool(toolName) {
+  return toolName === 'Agent' || toolName === 'agent';
+}
+
+function resolveSubagentTranscriptPath(parentTranscriptPath, agentId) {
+  const safeAgentId = path.basename(String(agentId || '')).replace(/\.jsonl$/, '');
```
Comment (ralf0131) [inline assets/hooks/claude-code-hook-processor.mjs]: **[Info]** The `path.basename` sanitization is good for preventing path traversal. One minor edge case: if `agentId` contains URL-encoded characters or unicode, `path.basename` will preserve them. Consider also validating that the resolved path stays within the expected `subagents/` directory (e.g. `resolvedPath.startsWith(expectedDir)`) as a defense-in-depth measure.

File: tests/unit/hooks/claude-code/skill-read.test.mjs [hunk 3]
```diff
@@ -558,8 +558,7 @@ describe('不变量: 同窗 Skill+注入 / offset 仅 Stop 推进 / 无 mid-turn
     const keys = [...dispatchBlock.matchAll(/'([^']+)':/g)].map((m) => m[1]).sort();
     expect(keys).toEqual(['stop', 'subagent-start', 'subagent-stop']);
 
-    // transcript 仅在(cmdStop 调用的)exportSession 里解析一次;offset 提升仅一处。
-    expect((src.match(/parseClaudeTranscript\(/g) || []).length).toBe(1);
+    // offset 提升仍仅允许一处。
```
Comment (ralf0131) [inline tests/unit/hooks/claude-code/skill-read.test.mjs]: **[Warning]** The PR removes the `parseClaudeTranscript` call-count assertion but does not add dedicated tests for the new subagent functionality. Consider adding unit tests for:
- `resolveSubagentTranscriptPath()` — especially path traversal edge cases (`../`, encoded paths)
- `collectSubagentLinks()` — with mock turns containing Agent tool calls
- The subagent export logic — verifying that child records inherit `trace_id`, `session.id`, `turn.id` and get the correct `gen_ai.agent.*` attributes

This would significantly improve confidence in the new code paths.

**Review by ralf0131 — COMMENTED** [review]
test

**Comment by ralf0131** [comment]
﻿### Summary
This PR adds first-level subagent trace reporting for Claude Code — correlating `Agent` tool results with their subagent transcripts, emitting proper spans for delegated LLM/tool calls, and inheriting parent trace context. The implementation is well-structured with good security (path traversal protection) and resilience (fail-open behavior). Two suggestions for improvement below.

### Findings
- **[Warning]** `claude-code-hook-processor.mjs:516` — Add try/catch around `parseClaudeTranscript()` for subagent transcripts to prevent malformed files from aborting the parent export
- **[Warning]** `skill-read.test.mjs:561` — Consider adding dedicated unit tests for the new subagent functionality (`resolveSubagentTranscriptPath`, `collectSubagentLinks`, subagent export logic)
- **[Info]** `claude-code-hook-processor.mjs:93` — Consider defense-in-depth path validation beyond `path.basename`

### Suggestions
1. Wrap the subagent transcript parsing in a try/catch to fully honor the fail-open contract:
   ```javascript
   try {
     const childParseResult = parseClaudeTranscript(childTranscriptPath, 0);
     // ... process child turns
   } catch (err) {
     console.warn(`[pilot] failed to parse subagent transcript ${childTranscriptPath}: ${err.message}`);
   }
   ```
2. Add unit tests for `resolveSubagentTranscriptPath()` covering edge cases (empty agentId, path traversal attempts, unicode characters).

### Positive Notes
- Clean separation of concerns with `isAgentTool()`, `resolveSubagentTranscriptPath()`, and `collectSubagentLinks()` helpers
- Proper path sanitization with `path.basename` and `.`/`..` checks
- Correct depth-limiting policy (first-level only) with clear documentation
- Good use of BigInt for nanosecond-precision timestamp sorting
- Backward compatible — no schema or API changes

---
*Automated review by github-manager-bot*

File: assets/hooks/claude-code-hook-processor.mjs (hunk 1 above)
Comment (zhouxiaohuan) [reply assets/hooks/claude-code-hook-processor.mjs]: Fixed in 85655a8. Subagent transcript parsing is now isolated in a per-child try/catch; failures are recorded through logHookError with stage=subagent_transcript_parse, then that child is skipped so the parent export and other children continue. A malformed-child regression test verifies the parent Agent tool record is still emitted.

File: tests/unit/hooks/claude-code/skill-read.test.mjs (hunk 3 above)
Comment (zhouxiaohuan) [reply tests/unit/hooks/claude-code/skill-read.test.mjs]: Added behavior-level coverage in 85655a8 through hook-processor.test.mjs, keeping the internal helpers private. The tests exercise Agent-link collection via mock turns, child export inheritance for trace_id/session.id/turn.id and gen_ai.agent attributes, malformed transcript fail-open behavior, traversal rejection, empty IDs, and Unicode IDs. The Claude Code hook suite passes 65 tests; the full suite passes 2108 tests with 5 skipped.

File: assets/hooks/claude-code-hook-processor.mjs (hunk 2 above)
Comment (zhouxiaohuan) [reply assets/hooks/claude-code-hook-processor.mjs]: Addressed in 85655a8. The resolver rejects empty IDs, separators, and NUL, resolves the candidate under the expected subagents directory, and verifies containment with path.relative. Unicode IDs remain supported. Behavior tests cover traversal, empty, and Unicode IDs.

**Review by ralf0131 — APPROVED** [review]
### Summary

All previous feedback has been addressed in commit 85655a8. The subagent transcript parsing is now wrapped in a per-child try/catch with proper error logging, comprehensive unit tests have been added (5 test cases covering happy path, fail-open, path traversal, empty ID, and unicode), and the path validation has been strengthened with defense-in-depth checks (empty ID, separators, NUL, containment via path.relative).

The implementation is clean, well-tested, and backward compatible. LGTM.

---
*Automated review by github-manager-bot*

**Comment by zhouxiaohuan** [comment]
Hi @ralf0131, I have merged the latest upstream main and resolved the conflicts in ce475b1.

Both the upstream Skill context changes and this PR's subagent trace reporting are preserved. Local verification passed:
- 58/58 Claude Code tests
- 2132 full-suite tests
- typecheck
- build

Could you please approve the CI workflow and re-review the latest commit? Thanks!

**Review by ralf0131 — APPROVED** [review]
### Summary

This PR adds first-level subagent trace reporting for Claude Code — correlating `Agent` tool calls with their generated subagent transcripts and exporting them into the parent turn trace with proper hierarchy attributes.

**Overall: LGTM.** Well-structured implementation with excellent attention to security and error handling.

### Highlights

- **Path traversal protection** — multi-layered validation in `resolveSubagentTranscriptPath()` (early rejection of `/`/`\\`/`\0`, `path.basename()` normalization, relative path containment check) provides robust defense-in-depth against malicious agent IDs
- **Graceful degradation** — per-child try/catch ensures a malformed subagent transcript cannot break the parent session export, with proper error logging at each failure point
- **Correct trace context inheritance** — subagent records properly inherit `trace_id`, `gen_ai.session.id`, and `gen_ai.turn.id` from the parent, with explicit `gen_ai.agent.depth=1` and `gen_ai.agent.scope=subagent` annotations
- **No more silent drops** — `Agent` tool calls now emit proper TOOL spans instead of being silently skipped
- **Comprehensive tests** — covers normal flow, malformed transcripts, path traversal attempts, empty/Unicode agent IDs

### Minor Observations (non-blocking)

1. **`resolveSubagentTranscriptPath`** — the early `/`/`\\` check makes the later `relativePath` containment check redundant for most cases, but the defense-in-depth is welcome and costs nothing
2. **First-level only** — the intentional non-recursive design (only expanding direct children) is correct for now; if deeper nesting is needed later, the architecture supports it cleanly via recursion on the same pattern

---
*Automated review by github-manager-bot*

**Comment by ralf0131** [comment]
Thanks for the update @zhouxiaohuan! 

I have re-reviewed the latest commit (`ce475b1`) and approved the PR. CI is all green — all checks passed.

The PR is ready for merge. A maintainer will pick it up shortly.

---
*Automated notification by github-manager-bot*

**Comment by ralf0131** [comment]
Thanks for the update @zhouxiaohuan! 

I have re-reviewed the latest commit (`ce475b1`) and approved the PR. CI is all green — all checks passed.

The PR is ready for merge. A maintainer will pick it up shortly.

---
*Automated notification by github-manager-bot*

**Review by fangxiu-wf — CHANGES_REQUESTED** [review]
@zhouxiaohuan Thank you very much for contributing this improvement. We are very happy to see contributions from external developers, and first-level Claude Code subagent observability is a valuable addition to the project.

I verified this PR against the real spans exported by the E2E run in AGE-1179. The foreground path is working: the `Agent` TOOL spans are correctly connected to child AGENT spans, token attributes are consistent, and no duplicate span IDs were observed.

However, two issues still need to be addressed before merge.

### 1. Background subagent traces are exported before the child transcript is complete

All 6 background-subagent launch traces were incomplete:

- none of the child traces contained the expected final output in `gen_ai.output.messages`;
- 5 of 6 child LLM chains ended at `tool_call`;
- the remaining one ended with `stop` but had zero output tokens and no final output;
- the later task-notification turns only contained a parent-agent summary and did not attach the completed child trace back to the original `Agent` span.

This matches the current implementation: the child transcript is read with `parseClaudeTranscript(childTranscriptPath, 0)` during the parent Stop export, without distinguishing an async launch from a completed Agent result or retaining state for completion-time export.

As a consequence, all 6 background traces also had an Agent TOOL span ending 1.991–4.038 seconds after its parent STEP ended. One trace contained a zero-duration Bash span without a tool result.

Please add lifecycle handling for background subagents so that their completed LLM/tool records are exported under the original Agent TOOL span without duplication.

### 2. The advertised hierarchy attributes are not present in the final exported spans

Across the 163 spans produced by this E2E run, the following attributes had zero occurrences:

- `gen_ai.agent.scope`
- `gen_ai.agent.depth`
- `gen_ai.agent.parent.id`
- `gen_ai.subagent.parent_tool_call.id`
- `gen_ai.turn.id`

`gen_ai.agent.id`, `gen_ai.agent.name`, and `gen_ai.session.id` were present, and the parent-child topology was correctly represented through `parentSpanId`.

The hook writes the missing fields into the intermediate records, but they are not preserved as attributes on the final spans. Since the PR description explicitly advertises these as hierarchy/context attributes, please either preserve them in the exported spans or clarify that they are internal converter fields and update the documented/tested contract accordingly.

The unit tests should also cover the async launch → completion-notification lifecycle and assert the final converted/exported spans, rather than only a completed transcript fixture or intermediate event records.

We look forward to the fixes. Once they are available, we will run the E2E validation again against the real exported trace data. If the validation passes, we will move forward with merging the PR as soon as possible. Thanks again for the contribution!

**Comment by ralf0131** [comment]
Thanks for resolving the conflicts! I have already approved this PR and the CI is passing. No further action needed from my side — the PR is ready for merge once the maintainers approve.

---
*Automated response by github-manager-bot*

**Comment by zhouxiaohuan** [comment]
Hi @fangxiu-wf, @ralf0131,

I have addressed the requested changes in commits `5458a94` and `e7a6997`:

- Background subagent turns are now deferred at the parent `Stop` and exported only after `SubagentStop`, using the completed child transcript.
- Completed child LLM/tool records are attached to the original `Agent` TOOL span without duplicate exports.
- Duplicate completion notifications and multiple background subagents in the same turn are handled idempotently.
- The five hierarchy attributes are now preserved on the final exported OTLP spans:
  - `gen_ai.agent.scope`
  - `gen_ai.agent.depth`
  - `gen_ai.agent.parent.id`
  - `gen_ai.subagent.parent_tool_call.id`
  - `gen_ai.turn.id`
- Tests now cover delayed completion, both event orderings, duplicate notifications, multiple background subagents, final output, span containment, hierarchy attributes, and unique span IDs.

Local verification:
- 89 focused tests passed
- 2227 full-suite tests passed, 6 skipped
- Typecheck and build passed

The latest CI run is currently `action_required`:
https://github.com/alibaba/loongsuite-pilot/actions/runs/30532848363

Could a maintainer please approve the workflow run?

@fangxiu-wf, once CI passes, could you please rerun the AGE-1179 E2E validation and re-review the latest head `e7a6997`?

Thank you!

**Review by ralf0131 — APPROVED** [review]
### Summary

Excellent feature addition — subagent trace merging with proper security hardening. The implementation correctly handles path traversal prevention, state locking, and background agent lifecycle.

### Review Highlights

- **Security**: `resolveSubagentTranscriptPath` has thorough path traversal protection (rejects `..`, `/`, `\`, null bytes, and validates relative path containment). Well done.
- **Concurrency safety**: `withStateLock` prevents race conditions during state mutations — critical for background agent scenarios.
- **Trace linking**: `collectSubagentLinks` correctly discovers subagent transcripts from tool call outputs and links them via `loongsuite.subagent.links`.
- **Backward compatibility**: Gracefully handles missing files and legacy state formats.
- **Tests**: Comprehensive coverage for path resolution edge cases, state locking, and transcript merging.

### Minor Suggestions

- Consider documenting the `FINALIZED_SUBAGENT_LIMIT = 128` constant with a comment explaining why 128 was chosen (memory vs. completeness tradeoff).
- The `resolveSubagentTranscriptPath` function could benefit from a JSDoc comment describing the expected directory structure.

Welcome to the project, @zhouxiaohuan! 🎉

---
*Automated review by github-manager-bot*

**Comment by zhouxiaohuan** [comment]
Hi @ralf0131, @fangxiu-wf,

The two minor documentation suggestions have been addressed in commit `86c09a0`:

- documented why `FINALIZED_SUBAGENT_LIMIT` keeps a bounded completion tombstone history;
- added JSDoc describing the expected direct-subagent transcript directory structure and safe-resolution behavior.

This follow-up contains comments only and does not change runtime behavior. The Claude Code hook test suite still passes: 25/25 tests.

Could a maintainer please approve the latest CI workflow run and re-review the new head `86c09a0`?

@fangxiu-wf, after CI passes, could you please rerun the AGE-1179 E2E validation against this head? The functional fixes remain in `5458a94` and `e7a6997`.

**Review by fangxiu-wf — APPROVED** [review]
@zhouxiaohuan Thank you again for the contribution and for addressing the review feedback. We reran the targeted E2E validation and independently verified the exported spans in SLS. The previously blocking issues are now resolved.

E2E evidence from AGE-1185:

- Source: PR merge result `b6b3349450ac141275553f9065dfe0303eee8530`, containing the functional fixes through `e7a6997`; the latest head `86c09a0` adds comments only and was separately verified with 40/40 targeted tests plus typecheck.
- Runtime: Claude Code 2.1.220 through the direct authenticated CLI path, with no test-side `NODE_OPTIONS`, `BUN_OPTIONS`, `BASH_ENV`, wrapper, or manually written telemetry.
- Foreground, one background subagent, and two concurrent background subagents all completed successfully.
- Original trace samples: foreground `416ef6d4715d4900b6935977f9f33856`, background `ae17b66933678ab3bc30691581c18df4`, concurrent background `956d6cd8e2533622bc88495082d87ae0`.
- Completed background child output and tool results are present under the original parent trace, rather than only in later task-notification traces.
- All four child AGENT spans are parented to the matching Agent TOOL spans. The five hierarchy attributes are present on the exported subagent spans: `gen_ai.turn.id`, `gen_ai.agent.scope`, `gen_ai.agent.depth`, `gen_ai.agent.parent.id`, and `gen_ai.subagent.parent_tool_call.id`.
- No parent/child time-containment violations, non-positive durations, duplicate span IDs, or duplicate child exports were found.
- Token coverage was 100% for the 86 eligible LLM spans, and the required message/tool data was present.

The E2E verdict is PASS. The run resources were cleaned up after evidence collection.

Approved from the code-review and E2E-acceptance perspective. Once the latest CI workflow is authorized and passes, this PR is ready to merge. Thank you for improving Claude Code subagent observability in LoongSuite Pilot.

**Review by ralf0131 — APPROVED** [review]
Re-review after documentation updates (commit 86c09a0).

Changes look solid:
- Path traversal protection in `resolveSubagentTranscriptPath` is thorough
- File-based state locking with stale lock recovery handles concurrent hook execution correctly
- `FINALIZED_SUBAGENT_LIMIT = 128` keeps tombstone state bounded
- Fail-open behavior for missing subagent transcripts is appropriate
- Test coverage is comprehensive (background subagents, duplicate events, path traversal, Unicode IDs)
- All CI checks passing

LGTM — approving.

---
*Automated review by github-manager-bot*
